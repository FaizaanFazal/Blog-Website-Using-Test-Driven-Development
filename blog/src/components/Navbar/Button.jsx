import React from 'react';

export default function Button(props) {
  const { className, onClick, children } = props;
  return (
    <button type="button" className={className} onClick={onClick}>
      {children}
    </button>
  );
}
