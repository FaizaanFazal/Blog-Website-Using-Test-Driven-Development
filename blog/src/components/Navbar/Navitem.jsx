import React from 'react'

export default function Navitem(props) {
  return (
    <li className="nav-item">
    <a href={props.href} className="nav-link">{props.text}{props.children}</a>
  </li>
  )
}
