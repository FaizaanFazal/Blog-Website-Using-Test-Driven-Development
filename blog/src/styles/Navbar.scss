@import "../App";

.nav-brand,
.ftr-brand {
  font-size: 24px;
  font-weight: bold;
  color: $color-white;

  &-dot {
    width: 6px;
    height: 6px;
    display: inline-block;
    border-radius: 100%;
    margin-left: 2px;
    color:$color-white ;
    background-color: $color-white;
   
  }
}

.navbar {
  height: 80px;
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  z-index: 999;

 

  .brand-and-toggler {
    @include breakpoint-down(lg) {
      width: 100%;
    }
  }

  .nav-show-btn {
    font-size: 24px;
    display: none;

    @include breakpoint-down(lg) {
      display: block;
    }
  }

  .nav-hide-btn {
    display: none;
    font-size: 24px;
  }

  .nb-menu-wrapper {
    .nav-item {
      margin-left: 40px;
    }

    .nav-link {
      color: $color-white;
      font-weight: 600;
      font-size: 16px;

      &-icon {
        display: inline-block;
        margin-left: 6px;
        margin-top: -2px;
      }
    }

    .nav-btns {
      margin-left: 40px;
    }

    .nav-btn {
      border: 2px solid $color-white;
      border-radius: 8px;
      color: $color-white;
      font-weight: 600;
      font-size: 16px;

      &:hover {
        background-color: $color-primary;
        color: $color-white;
        border-color: $color-primary;
      }
    }

    @include breakpoint-down(lg) {
      position: fixed;
      right: 0;
      top: 0;
      width: 240px;
      background-color: $color-white;
      height: 100%;
      z-index: 999;
      box-shadow: rgb(0 0 0 / 35%) 0 5px 15px;
      flex-direction: column;
      padding: 60px 16px;
      row-gap: 20px;
      transform: translateX(100%);
      transition: $default-transition;

      &.show {
        transform: translateX(0);
      }

      .nav-hide-btn {
        display: block;
        position: absolute;
        right: 16px;
        top: 6px;
      }

      .nav-menu {
        flex-direction: column;
        width: 100%;
      }

      .nav-item {
        margin-left: 0;
        padding-bottom: 16px;
        padding-top: 16px;
        border-bottom: 1px dashed rgb(0 0 0 / 20%);
        width: 100%;
        display: flex;
        justify-content: center;

        &:last-child {
          border-bottom: 0;
        }
      }

      .nav-link {
        color: $color-blue;
      }

      .nav-btns {
        margin-left: 0;
      }

      .nav-btn {
        border-color: $color-primary;
        color: $color-primary;

        &:hover {
          background-color: $color-primary !important;
          color: $color-white !important;
        }
      }
    }

    @include breakpoint-down(xs) {
      width: 100%;
    }
  }

  &.scrolled {
    background-color: $color-white;
    box-shadow: rgb(50 50 93 / 15%) 0 13px 27px -5px, rgb(0 0 0 / 10%) 0 8px 16px -8px;

    .nav-brand,
    .nav-link,
    .nav-show-btn {
      color: $color-blue;
    }

    .nav-brand {
      &-dot {
        background-color: $color-blue;
        color: $color-blue ;
      }
    }

    .nav-btn {
      border-color: $color-blue;
      color: $color-blue;

      &:hover {
        background-color: $color-blue;
        color: $color-white;
        border-color: $color-blue;
      }
    }
  }

  @include breakpoint-down(xs) {
    height: 64px;
  }
}

.animated-button {
  animation: move-down 0.5s linear normal;
}

@keyframes move-down {
  0% {
    opacity: 0;
    transform: translateY(-20px);
  }
  100% {
    opacity: 1;
    transform: translateY(0px);
  }
}
