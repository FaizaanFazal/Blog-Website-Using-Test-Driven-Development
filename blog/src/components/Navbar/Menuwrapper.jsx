import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import Navitem from './Navitem';
import { useSelector } from 'react-redux';

export default function Menuwrapper() {
  const [user, setUser] = useState();
  const userr = useSelector(state => state.user.user); //userName
  console.log(userr)

  useEffect(() => {
    if (userr["userName"] && userr["userName"].length > 0) {
      setUser(userr);
    }
    else {
      setUser(null)
    }
  }, [userr])

  return (
    <div>
      <ul className="nav-menu flex items-center">
        <Navitem to="/" text="Home" />
        {/* <Navitem to="/featured" text="Featured">
          <span className="nav-link-icon"><i className="bi bi-chevron-down" /></span>
        </Navitem> */}
        <Navitem to="/about" text="About" />
        <Navitem to="/contacts" text="Contact" />
        <div className="nav-btns">
          {user ? (
            <div className="nav-btn btn">{user.userName}</div>
          ) : (
            <Link data-testid="listitem" to="/login" className="nav-btn btn">
              Login
            </Link>
          )}
        </div>
        <div className="nav-btns">
          <Link data-testid="listitem" to="/createpost" className="nav-btn btn">Post</Link>
        </div>
      </ul>
    </div>
  );
}
