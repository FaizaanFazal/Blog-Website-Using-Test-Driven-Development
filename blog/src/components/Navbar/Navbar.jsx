import { useEffect, useState } from 'react';
import '../../styles/Navbar.scss';
import { Link } from 'react-router-dom';
import Menuwrapper from './Menuwrapper';
import Button from './Button';

const Navbar = () => {
  const [scrolled, setScrolled] = useState(false);
  const [isSideMenuOpen, setSideMenuOpen] = useState(false);

  const hideSideMenu = () => setSideMenuOpen(false);
  const showSideMenu = () => setSideMenuOpen(true);

  const handleScroll = () => {
    const offset = window.scrollY;
    if (offset > 60) setScrolled(true);
    else setScrolled(false);
  };

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
  }, []);

  return (
    <nav role="navigation" className={`navbar flex items-center ${scrolled ? 'scrolled' : ''}`}>
      <div className="container navbar-container w-full flex justify-between">
        <div className="brand-and-toggler flex items-center justify-between">
          <Link data-testid="logo" to="/" className="nav-brand">
            Blogs
            {' '}
            <span className="nav-brand-dot" />
          </Link>
          <Button className="nav-show-btn text-white" onClick={showSideMenu}><i className="bi bi-list" /></Button>
        </div>

        <div className={`nb-menu-wrapper flex items-center ${isSideMenuOpen ? 'show' : ''}`}>
          <Button className="nav-hide-btn" onClick={hideSideMenu}><i className="bi bi-x-square" /></Button>
          <Menuwrapper />
        </div>

      </div>
    </nav>
  );
};

export default Navbar;
