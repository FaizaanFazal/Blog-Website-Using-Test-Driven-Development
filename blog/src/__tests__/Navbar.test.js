import { render, screen,fireEvent } from '@testing-library/react';
import Navbar from '../components/Navbar/Navbar';

describe("",()=>{
  it('renders the navbar', () => {
    render(<Navbar />);
    const navbarElement = screen.getByRole('navigation');
    expect(navbarElement).toBeInTheDocument();
  });

  it('renders the logo', () => {
    render(<Navbar />);
    const logoElement = screen.getByTestId('logo');
    expect(logoElement).toBeInTheDocument();
  });

  it('checks list length', () => {
    render(<Navbar />);
    const linksElement = screen.getAllByRole('listitem');
    expect(linksElement.length).toBe(5); //home,featured,about,contact,createpost
  });

  //links check
  it('navigates to home page', () => {
    render(<Navbar />);
    const homeLink = screen.getByText('Home');
    fireEvent.click(homeLink);
    expect(window.location.pathname).toBe('/');
  });

  it('navigates to about page', () => {
    render(<Navbar />);
    expect(screen.getByText('About')).toHaveAttribute('href', '/about')
  });
  it('navigates to about page', () => {
    render(<Navbar />);
    expect(screen.getByText('Featured')).toHaveAttribute('href', '/featured')
  });

  it('navigates to contact page', () => {
    render(<Navbar />);
    expect(screen.getByText('Contact')).toHaveAttribute('href', '/contact')
  });

  it('navigates to contact page', () => {
    render(<Navbar />);
    expect(screen.getByText('Post')).toHaveAttribute('href', '/createpost')
  });

})