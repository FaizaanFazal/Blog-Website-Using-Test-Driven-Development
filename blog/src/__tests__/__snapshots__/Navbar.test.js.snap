// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for navbar snapshot test to check it renders Navbar component 1`] = `
<DocumentFragment>
  <nav
    class="navbar flex items-center "
    role="navigation"
  >
    <div
      class="container navbar-container w-full flex justify-between"
    >
      <div
        class="brand-and-toggler flex items-center justify-between"
      >
        <a
          class="nav-brand"
          data-testid="logo"
          href="/"
        >
          Blogs 
          <span
            class="nav-brand-dot"
          />
        </a>
        <button
          class="nav-show-btn text-white"
          type="button"
        >
          <i
            class="bi bi-list"
          />
        </button>
      </div>
      <div
        class="nb-menu-wrapper flex items-center "
      >
        <button
          class="nav-hide-btn"
          type="button"
        >
          <i
            class="bi bi-x-square"
          />
        </button>
        <div>
          <ul
            class="nav-menu flex items-center"
          >
            <li
              class="nav-item"
              data-testid="listitem"
            >
              <a
                class="nav-link"
                href="/"
                role="link"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
              data-testid="listitem"
            >
              <a
                class="nav-link"
                href="/about"
                role="link"
              >
                About
              </a>
            </li>
            <li
              class="nav-item"
              data-testid="listitem"
            >
              <a
                class="nav-link"
                href="/contacts"
                role="link"
              >
                Contact
              </a>
            </li>
            <div
              class="nav-btns"
            >
              <a
                class="nav-btn btn"
                data-testid="listitem"
                href="/login"
              >
                Login
              </a>
            </div>
            <div
              class="nav-btns"
            >
              <a
                class="nav-btn btn"
                data-testid="listitem"
                href="/createpost"
              >
                Post
              </a>
            </div>
          </ul>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
